### PR TITLE
Glaxnimate: init at 0.5.1

### LINKED ISSUE #208707: glaxnimate
**Project description**

simple and fast vector graphics animation program. glaxnimate works with vector graphics, this means images are described with objects like lines, curves, and points.

img

**Metadata**

* homepage URL: https://glaxnimate.mattbas.org
* source URL: https://gitlab.com/mattbas/glaxnimate/-/releases
* license: GPLv3+ <https://gitlab.com/mattbas/glaxnimate#license>
* platforms: unix, linux, darwin

### PR DESCRIPTION
**Pull request by tobiasBora**
###### Description of changes

Fix https://github.com/NixOS/nixpkgs/issues/208707#issuecomment-1374766082

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/video/glaxnimate/default.nix
Comment (wegank) [inline pkgs/applications/video/glaxnimate/default.nix:32]:
```suggestion
 
```

File: pkgs/applications/video/glaxnimate/default.nix
Comment (wegank) [inline pkgs/applications/video/glaxnimate/default.nix:13]:
```suggestion
```

File: pkgs/applications/video/glaxnimate/default.nix
Comment (wegank) [inline pkgs/applications/video/glaxnimate/default.nix:32]: Oh, my bad...

```suggestion

```

**Comment by turion** [comment]
I can build it and create a basic animation.

## Working

* Creating a basic animation
* SVG export
* Lottie export
* MP4 export

## Failing

* GIF export: `"2023-01-08T12:42:47" "Error" "Animated Raster" "" "Plugin raised an exception"`

I'd be fine with accepting this (as soon as whitespace issues are fixed) and opening the failing GIF export as a separate issue.

**Comment by tobiasBora** [comment]
Thanks a lot! Have you tried features requiring python?

**Comment by turion** [comment]
Some simple ideas for an integration test in `nixos/tests`:

* Check the output of `${glaxnimate}/bin/glaxnimate --version`
* Start `glaxnimate` and then use `machine.wait_for_window` and `machine.wait_for_text` to ensure the main window comes up

**Comment by turion** [comment]
(But an integration test wouldn't be strictly necessary.)

**Comment by turion** [comment]
> Have you tried features requiring python?

No, I'm not sure which these are.

File: pkgs/applications/video/glaxnimate/default.nix
Comment (turion) [inline pkgs/applications/video/glaxnimate/default.nix:47]: Sure it doesn't build on darwin? There are installation instructions for macOS: https://gitlab.com/mattbas/glaxnimate#macos Maybe it works with some simple darwin-specific CMAKE flags?

Comment (tobiasBora) [reply pkgs/applications/video/glaxnimate/default.nix:47]: I've no ideas, but as I can't test on darwin I don't know what I should put. Maybe someone from @NixOS/darwin-maintainers can test? Btw @turion do you want to be part of tge maintainers as you certainly use this software more than me?

**Comment by tobiasBora** [comment]
> No, I'm not sure which these are.

Python seems to be useful for scripting and provides a console.  I might give it a try when I have more time in the coming week https://glaxnimate.mattbas.org/contributing/scripting/

File: pkgs/applications/video/glaxnimate/default.nix
Comment (turion) [reply pkgs/applications/video/glaxnimate/default.nix:47]: You could add a simple integration test, then ofborg can test it.

I wouldn't want to be maintainer. I use kdenlive, but have never used glaxnimate, nor do I intend to.

Comment (Et7f3) [reply pkgs/applications/video/glaxnimate/default.nix:47]: I am building it on x86-64_darwin

File: pkgs/applications/video/glaxnimate/default.nix
Comment (Et7f3) [inline pkgs/applications/video/glaxnimate/default.nix:37]: it is a binary so nativeBuildInputs ? same for qttools ?

File: pkgs/applications/video/glaxnimate/default.nix
Comment (Et7f3) [reply pkgs/applications/video/glaxnimate/default.nix:47]: seem fine 
image

File: pkgs/applications/video/glaxnimate/default.nix
Comment (turion) [reply pkgs/applications/video/glaxnimate/default.nix:37]: I'd imagine that `python3` is needed at runtime for scripting.

**Comment by turion** [comment]
> and provides a console + python library.

For that, a separate package under the python namespace would be useful. (But not necessary for this PR)

File: pkgs/applications/video/glaxnimate/default.nix
Comment (tricktron) [inline pkgs/applications/video/glaxnimate/default.nix:47]:
```suggestion

```
It also builds on aarch64-darwin, so you can remove that line which means it works on the default platforms.

**Comment by tricktron** [comment]
> Some simple ideas for an integration test in `nixos/tests`:
> 
> * Check the output of `${glaxnimate}/bin/glaxnimate --version`

I tried to add this test with:

```nix
passthru.tests.version = testers.testVersion {
    package = glaxnimate;
    command = "glaxnimate --version";
  };
```

However, on aarch64-darwin running this with `nix-build -A glaxnimate.passthru.tests.version` fails with `glaxnimate --version returned a non-zero exit code.`

On the other hand, building glaxnimate and the running manually `result/bin/glaxnimate --version` works.

Any ideas?

**Comment by turion** [comment]
I have no system to reproduce right now, but you could push a commit containing the test and we can look at the ofborg output.

File: pkgs/applications/video/glaxnimate/default.nix
Comment (turion) [inline pkgs/applications/video/glaxnimate/default.nix:35]:
```suggestion
    # Upstream asks for libav dependency, which is fulfilled by ffmpeg
    ffmpeg
```
(Or a more aptly worded comment)

**Review by wegank — APPROVED** [review]
Tested on aarch64-darwin.

File: pkgs/applications/video/glaxnimate/default.nix
Comment (tricktron) [inline pkgs/applications/video/glaxnimate/default.nix:12]:
```suggestion
, wrapQtAppsHook
, testers
, glaxnimate
```

File: pkgs/applications/video/glaxnimate/default.nix
Comment (tricktron) [inline pkgs/applications/video/glaxnimate/default.nix:41]:
```suggestion

passthru.tests.version = testers.testVersion {
    package = glaxnimate;
    command = "glaxnimate --version";
  };
```

**Review by tricktron — COMMENTED** [review]
Use `glaxnimate --version` as integration test. However, this currently fails on aarch64-darwin and possibly on linux too. If you add this, we can look at ofborg to investigate the issue.

**Comment by tobiasBora** [comment]
Thanks. I think I understand the issue for Gif export, PIL is missing from python, with possibly other libraries. I'll investigate tomorrow.

**Comment by tricktron** [comment]
@NixOS/ofborg build glaxnimate glaxnimate.passthru.tests.version

File: pkgs/applications/video/glaxnimate/default.nix
Comment (Et7f3) [reply pkgs/applications/video/glaxnimate/default.nix]: Does it compile on windows ? if no add `platforms = platforms.unix;` (try to cross compile with mingw)

Comment (Et7f3) [reply pkgs/applications/video/glaxnimate/default.nix]: where is default platforms list ? freebsd is inside ?

Comment (wegank) [reply pkgs/applications/video/glaxnimate/default.nix]: I guess the MinGW cross chain is mostly broken though...

Comment (wegank) [reply pkgs/applications/video/glaxnimate/default.nix]: @Et7f3 Oh, you really made my day!

**Comment by tobiasBora** [comment]
The plugins & gif export issues are solved… the problem of `glaxnimate.passthru.tests.version` is that it tries to start the display, and seems like it's not available in these tests.
```
glaxnimate --version returned a non-zero exit code.
qt.qpa.xcb: could not connect to display 
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```
Not sure if there is a way to provide a display in that case.

**Comment by tobiasBora** [comment]
Ok, I fixed the test script using `xvfb-run`. Hopefully everything should work now, feel free to check a last time so that we can merge it.

File: pkgs/applications/video/glaxnimate/default.nix
Comment (tobiasBora) [reply pkgs/applications/video/glaxnimate/default.nix]: Yeah, I agree with @turion , these are libraries/binaries used at runtime so I see no reasons why to put them in `nativeBuildInputs` that should only contain stuff needed to configure the build process (see some files provided by qttools here for instance https://github.com/qt/qttools/blob/dev/src/qtdiag/qtdiag.cpp)

**Comment by turion** [comment]
Awesome, thanks a lot!

**Comment by winterqt** [comment]
@turion In the case where fixes are done in separate commits, can you please squash and merge next time? 10 out of these 11 commits are just cluttering up the history now 🙃 thanks! (Luckily it's not too big of a deal since you rebased them, so I think that should avoid them being spread out...)

**Comment by turion** [comment]
Sorry, I forgot to check the commit history! I'll try and remember the next times.

**Comment by tobiasBora** [comment]
Yeah, I was expecting the squash to be done at merge time, hence the dirty history, I should maybe add a warning next time to remind to squash.

**Comment by tricktron** [comment]
@turion @tobiasBora
xvbf is not supported on darwin. So running `nix-build -A glaxnimate.passthru.tests.version` will always fail on darwin with:
```console
Package ‘xvfb-run’ in /Users/tricktron/github/my-forks/nixpkgs/pkgs/tools/misc/xvfb-run/default.nix:59 is not supported on ‘aarch64-darwin’, refusing to evaluate.
```

I don't know any replacement for xvfb-run on darwin and since the testers.testVersion function is mostly intended for binaries without any GUI I think it is best to just disable the passthru test on darwin.

Maybe something like: 
```nix
passthru.tests.version = lib.optionalAttrs stdenv.isLinux (testers.testVersion {
    package = glaxnimate;
    command = "${xvfb-run}/bin/xvfb-run glaxnimate --version";
  });
```

**Comment by tobiasBora** [comment]
Oh good to know. Sure, feel free to write a PR !

**Comment by turion** [comment]
Weird, all the CI checks were green. Aren't the tests run automatically?

**Comment by winterqt** [comment]
> Weird, all the CI checks were green. Aren't the tests run automatically?

@turion Normally, yes. But in this case, the commit message is wrong (should be lowercase to match the attr name, which is how OfBorg knows what to build automatically), so no. I'm admittedly not sure if it builds if more commits are added on top of that that *don't* include that name, though.

**Comment by turion** [comment]
Ok, lesson certainly learned: Inspect commit messages closely before merging.

**Pull request merged**
